cpp/generate_title_and_author_list.cc

```diff
@@ -23,50 +23,42 @@
 #include <cstdio>
 #include <cstdlib>
 #include "Compiler.h"
-#include "MarcReader.h"
-#include "MarcRecord.h"
-#include "Subfields.h"
+#include "MARC.h"
 #include "util.h"
 
 
-static void Usage() __attribute__((noreturn));
+namespace {
 
 
-static void Usage() {
+[[noreturn]] void Usage() {
     std::cerr << "Usage: " << ::progname << " marc_data\n";
     std::exit(EXIT_FAILURE);
 }
 
 
-void ExtractAuthors(const MarcRecord &record, std::vector<std::string> * const authors) {
+void ExtractAuthors(const MARC::Record &record, std::vector<std::string> * const authors) {
     authors->clear();
 
-    std::vector<size_t> field_indices;
-    record.getFieldIndices("100", &field_indices);
-    for (const size_t field_index : field_indices) {
-        const std::string field_contents(record.getFieldData(field_index));
-        if (unlikely(field_contents.empty()))
-            continue;
-
-        const Subfields subfields(field_contents);
-        const std::string author(subfields.getFirstSubfieldValue('a'));
+    for (const auto &field : record.getTagRange("100")) {
+        const auto subfields(field.getSubfields());
+        const std::string author(subfields.getFirstSubfieldWithCode('a'));
         if (likely(not author.empty()))
             authors->emplace_back(author);
     }
 }
 
 
-void ProcessRecords(MarcReader * const marc_reader) {
+void ProcessRecords(MARC::Reader * const marc_reader) {
     unsigned record_count(0);
-    while (const MarcRecord record = marc_reader->read()) {
+    while (const MARC::Record record = marc_reader->read()) {
         ++record_count;
 
-        const std::string _245_contents(record.getFieldData("245"));
-        if (unlikely(_245_contents.empty()))
+        const auto field_245(record.findTag("245"));
+        if (unlikely(field_245 == record.end()))
             continue;
 
-        const Subfields subfields(_245_contents);
-        const std::string main_title(subfields.getFirstSubfieldValue('a'));
+        const auto subfields(field_245->getSubfields());
+        const std::string main_title(subfields.getFirstSubfieldWithCode('a'));
```

Comment:
I don't think we need "subfields".  I think you can use field_245->getFirstSubfieldWithCode('a') directly.